internal/dag/gatewayapi_processor.go

```diff
@@ -83,6 +91,43 @@ func (p *GatewayAPIProcessor) Run(dag *DAG, source *KubernetesCache) {
 
 }
 
+// namespaceMatches returns true if the namespaces selector matches
+// the HTTPRoute that is being processed.
+func (p *GatewayAPIProcessor) namespaceMatches(namespaces gatewayapi_v1alpha1.RouteNamespaces, route *gatewayapi_v1alpha1.HTTPRoute) (bool, error) {
+	// From indicates where Routes will be selected for this Gateway.
+	// Possible values are:
+	//   * All: Routes in all namespaces may be used by this Gateway.
+	//   * Selector: Routes in namespaces selected by the selector may be used by
+	//     this Gateway.
+	//   * Same: Only Routes in the same namespace may be used by this Gateway.
+
+	switch namespaces.From {
+	case gatewayapi_v1alpha1.RouteSelectAll:
+		return true, nil
+	case gatewayapi_v1alpha1.RouteSelectSame:
+		return p.source.Gateway.Namespace == route.Namespace, nil
+	case gatewayapi_v1alpha1.RouteSelectSelector:
+		if len(namespaces.Selector.MatchLabels) == 0 || len(namespaces.Selector.MatchExpressions) == 0 {
+			return false, fmt.Errorf("RouteNamespaces selector must be specified when `RouteSelectType=Selector`")
+		}
+
+		// Look up the HTTPRoute's namespace in the list of cached namespaces.
+		if ns := p.source.namespaces[types.NamespacedName{Name: route.Namespace, Namespace: "default"}]; ns != nil {
```

Comment:
The fact that the key here has a `Namespace: "default"` is confusing because `Namespaces` are cluster-scoped. I'm seeing that it's due to `k8s.NamespacedNameOf(...)` automatically setting a namespace of "default" if it's not provided. I'm thinking we should change the `namespaces` map in the DAG cache to be keyed just on a string (the name), rather than a `NamespacedName`, and not use `NamespacedNameOf` when getting/setting. This looks like the pattern that `IngressClass` uses, which is also cluster-scoped.

Reply:
Yup good call! It didn't feel right when I did that, but this makes more sense. I opened https://github.com/projectcontour/contour/pull/3473 to cover.

Reply:
There is some discussion upstream about making both GatewayClass and IngressClass either namespace-scoped, or have namespace-scoped variants.